Feature/first view controller

VikkiRy opened the pull request.

VikkiRy committed:
Data loaded

VikkiRy committed:
Unused code removed.
setupUI() updated.

VikkiRy committed:
- Combine ServiceAssemblyProtocol and NewsServiceAssemblyProtocol.
- Combine ViewModelAssemblyProtocol and NewsViewModelAssembly.
- Code refactored.

nikGayko reviewed (changes requested):
testing, don't pay attention

nikGayko commented on BadHabbit/BadHabbit.xcodeproj/project.pbxproj:
```diff
@@ -7,6 +7,14 @@
 	objects = {
 
 /* Begin PBXBuildFile section */
+		61BA170A2A1F84C500CB7FAD /* NewsScreenAssembly.swift in Sources */ = {isa = PBXBuildFile; fileRef = 61BA17092A1F84C500CB7FAD /* NewsScreenAssembly.swift */; };
```
I don't like it. Because of this and that.

VikkiRy committed:
- Extra line code removed.
- ArticleResponse and Article classes are refactored.
- requestExecutor.perform refactored.

VikkiRy committed:
TextLabel transformed to TableView

nikGayko commented on BadHabbit/BadHabbit/Kit/NetworkKit/ProtocolType.swift:
```diff
@@ -0,0 +1,12 @@
+
+import Foundation
+
+enum ProtocolType: String {
```
Confusing name. Protocol we also use to declare interfaces. Try to come out with more suitable name

VikkiRy replied:
Resolved

nikGayko commented on BadHabbit/BadHabbit/Kit/NetworkKit/Request.swift:
```diff
@@ -5,8 +5,38 @@
+        
+        let queryItems = urlQueryItems()
+        
+        guard !queryItems.isEmpty else {
```
Logic too complicated. Could be more simple. If we replace with `it` statement and keep only one return from function

VikkiRy replied:
Resolved

nikGayko commented on BadHabbit/BadHabbit/Kit/NetworkKit/RequestExecutor.swift:
```diff
@@ -27,16 +28,37 @@ class RequestExecutor: RequestExecutorProtocol {
         }
+        
+        URLSession.shared.dataTask(with: url) { data, _, error in
+            guard let data = data else {
```
what if data is nil but error is still presented ?

nikGayko replied:
Does it better to return `error` passed into this closure or generic `dataError` ?

VikkiRy replied:
I assumed here: _no data and no error_, but my code was wrong

nikGayko replied:
For now if there there is error and data is nil you call completion twice - this isn't correct. Completion means that it called when function complete it's async execution.

VikkiRy replied:
Bug fixed (keyword return added)

Resolved

VikkiRy replied:
In the latest update, `if error = nil && response.statuscode < 400`, the data will always have a value.
In our implementation, `if data.isEmpty {...}` we will throw a NetworkError.parseError

Resolved

nikGayko commented on BadHabbit/BadHabbit/SceneDelegate.swift:
```diff
@@ -33,6 +38,7 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
         
         let window = UIWindow(windowScene: windowScene)
         window.rootViewController = rootTabBarAssembler.assemble()
+        window.backgroundColor = .white
```
why ?

VikkiRy replied:
This line of code was removed, tabBarItems color was changed.

Resolved

nikGayko commented on BadHabbit/BadHabbit/UserStories/NewsScreen/NewsScreenViewController.swift:
```diff
@@ -0,0 +1,122 @@
+    private let viewModel: ExampleViewModelProtocol
+    
+    let textLabel = UILabel()
+    let errorLabel = UILabel()
```
should be private

VikkiRy replied:
Resolved

nikGayko commented on BadHabbit/BadHabbit/UserStories/NewsScreen/NewsScreenViewController.swift:
```diff
@@ -0,0 +1,122 @@
+        loadDataButton.backgroundColor = .systemGreen
+        
+        loadDataButton.addTarget(self, action: #selector(loadButtonTapped), for: .touchUpInside)
+        loadDataButton.tag = 1
```
what is tag for ?

nikGayko replied:
If you need to access this view out of this function - just make property not local but class property

VikkiRy replied:
Resolved

nikGayko commented on BadHabbit/BadHabbit/UserStories/NewsScreen/NewsScreenViewController.swift:
```diff
@@ -0,0 +1,122 @@
+    
+    private func bindVM() {
+        viewModel.exampleText = { [weak self] text in
+            DispatchQueue.main.async {
```
this calls must be in viewModel

VikkiRy replied:
Closure logic changed.

Resolved

nikGayko replied:
Queue changing should be performed in VM, not in VC. In the same way as UIKit elements like UILabel and UITextField doesn't check what queue you use to change them, you also shouldn't do that. You should transition execution to mainQueue in UI layer

VikkiRy replied:
Resolved

nikGayko commented on BadHabbit/BadHabbit/UserStories/NewsScreen/NewsScreenViewController.swift:
```diff
@@ -0,0 +1,122 @@
+        NSLayoutConstraint.activate([
+            loadDataButton.rightAnchor.constraint(equalTo: view.rightAnchor, constant: -15.0),
+            loadDataButton.leftAnchor.constraint(equalTo: view.leftAnchor, constant: 15.0),
+            loadDataButton.heightAnchor.constraint(equalTo: view.heightAnchor, multiplier: 0.06, constant: 0),
```
Rarely you could find that button has height related to screen height

VikkiRy replied:
Resolved

VikkiRy committed:
- urlComponents() logic changed
- URLSession.shared.dataTask(with: url) { ... } logic changed
- enum ProtocolType name changed into UrlScheme
- tabBarItem tint color changed
- NewsScreenViewController refactored

VikkiRy committed:
- NewsViewModel class property name changed.
- loadData() refactored.
- loadButtonTapped() refactored.
- bindVM() property name changed.

VikkiRy committed:
Bug in if let error = error { ... } fixed

nikGayko commented on BadHabbit/BadHabbit/UserStories/NewsScreen/NewsScreenViewController.swift:
```diff
@@ -81,8 +81,8 @@ extension NewsScreenViewController {
     private func loadButtonTapped() {
-        DispatchQueue.main.async {
-            self.loadDataButton.removeFromSuperview()
+        DispatchQueue.main.async { [weak self] in
```
Could you explain why we need to capture weakly self here ? And why not needed ?

VikkiRy replied:
We need to define a reference type because async function captures `self` strongly. But here we should use an unowned reference because `self` will never become nil in our implementation.
Is it correct?

nikGayko replied:
What is the problem with capturing strongly self here ?

VikkiRy replied:
If we capture `self` strongly and we don't have any other references to the current ViewController instance, this instance won't be deinitialized and ARC won't deallocate it from memory until the closure isn't executed.

nikGayko replied:
Is there any chance that this closure won't be executed ?

VikkiRy replied:
Yes, there is. It will happen if our app is force closed (by user or by the system for some reason) before the closure will start execute.

nikGayko replied:
In case when app closed all memory allocated for this app is freed. Including memory allocated for this main queue and it DispatchOperations

nikGayko replied:
Are there other cases ?

VikkiRy replied:
In our case we don't have a strong reference cycle, so I removed the closure capture list.

VikkiRy committed:
- URLSession.shared.dataTask(with: url) refactored
- var response completion logic changed

VikkiRy committed:
bindVM() func changed

VikkiRy committed:
loadButtonTapped() function refactored
